src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.Async.cs

```diff
@@ -0,0 +1,608 @@
+                        stream.Seek(0, SeekOrigin.Begin);
+                    }
+                    break;
+                case ZipArchiveMode.Update:
+                    if (!stream.CanRead || !stream.CanWrite || !stream.CanSeek)
+                        throw new ArgumentException(SR.UpdateModeCapabilities);
+                    break;
+                default:
+                    // still have to throw this, because stream constructor doesn't do mode argument checks
+                    throw new ArgumentOutOfRangeException(nameof(mode));
+            }
+
+            Stream archiveStream;
+            if (mode == ZipArchiveMode.Create && !stream.CanSeek)
+            {
+                archiveStream = new PositionPreservingWriteOnlyStreamWrapper(stream);
+            }
+            else
+            {
+                archiveStream = stream;
+            }
+
+            ZipArchive zipArchive = new(mode, leaveOpen, entryNameEncoding, backingStream, archiveStream, cancellationToken);
+
+            switch (mode)
+            {
+                case ZipArchiveMode.Create:
+                    zipArchive._readEntries = true;
+                    break;
+                case ZipArchiveMode.Read:
+                    await zipArchive.ReadEndOfCentralDirectoryAsync(cancellationToken).ConfigureAwait(false);
+                    break;
+                case ZipArchiveMode.Update:
+                default:
+                    Debug.Assert(mode == ZipArchiveMode.Update);
+                    if (zipArchive._archiveStream.Length == 0)
+                    {
+                        zipArchive._readEntries = true;
+                    }
+                    else
+                    {
+                        await zipArchive.ReadEndOfCentralDirectoryAsync(cancellationToken).ConfigureAwait(false);
+                        await zipArchive.EnsureCentralDirectoryReadAsync(cancellationToken).ConfigureAwait(false);
+                        foreach (ZipArchiveEntry entry in zipArchive._entries)
+                        {
+                            entry.ThrowIfNotOpenable(needToUncompress: false, needToLoadIntoMemory: true);
+                        }
+                    }
+                    break;
+            }
+
+            return zipArchive;
+
+        }
+        catch
+        {
+            if (extraTempStream != null)
+            {
+                await extraTempStream.DisposeAsync().ConfigureAwait(false);
+            }
+
+            throw;
+        }
+    }
+
+    /// <summary>
+    /// This is a helper constructor for <see cref="CreateAsync"/>. It initializes most of the essential
+    /// ZipArchive information but the collection of entries needs to be done asynchronously, so it is
+    /// done by <see cref="CreateAsync"/>.
+    /// </summary>
+    private ZipArchive(ZipArchiveMode mode, bool leaveOpen, Encoding? entryNameEncoding, Stream? backingStream, Stream archiveStream, CancellationToken cancellationToken)
```

Comment:
nit: any value in cascading other constructors to this one so share this initialization?

Reply:
Interesting suggestion. Let me check.

Reply:
I checked, and no, I can't call this new private constructor from the old constructor because of the order in which things are verified and initialized. But I can make the constructors share more portions of code instead.

Reply:
On a second thought, I could try changing the order of some things that aren't risky.